services/statestorage/service.go

```diff
@@ -18,29 +18,38 @@ var LogTag = log.Service("state-storage")
 
 type service struct {
 	config       config.StateStorageConfig
-	merkle       *merkle.Forest
 	blockTracker *synchronization.BlockTracker
 	logger       log.BasicLogger
 
-	mutex                    *sync.RWMutex
-	persistence              adapter.StatePersistence
-	lastCommittedBlockHeader *protocol.ResultsBlockHeader
+	mutex        sync.RWMutex
+	layeredState *layeredState
+	merkle       *merkle.Forest
+	height       primitives.BlockHeight
+	ts           primitives.TimestampNano
 }
 
 func NewStateStorage(config config.StateStorageConfig, persistence adapter.StatePersistence, logger log.BasicLogger) services.StateStorage {
-	merkle, rootHash := merkle.NewForest()
-	// TODO this is equivalent of genesis block deploy in persistence -> move to correct deploy
-	persistence.WriteMerkleRoot(0, rootHash)
+	// TODO - tie/sync merkle forest to persistent state
+	merkle, root := merkle.NewForest()
+
+	height, ts, pRoot, err := persistence.ReadMetadata()
+	if err != nil {
+		panic(err)
+	}
+	if !pRoot.Equal(root) {
+		panic("Merkle forest out of sync with persisted state")
+	}
 
 	return &service{
 		config:       config,
-		merkle:       merkle,
 		blockTracker: synchronization.NewBlockTracker(0, uint16(config.BlockTrackerGraceDistance()), config.BlockTrackerGraceTimeout()),
 		logger:       logger.WithTags(LogTag),
 
-		mutex:                    &sync.RWMutex{},
-		persistence:              persistence,
-		lastCommittedBlockHeader: (&protocol.ResultsBlockHeaderBuilder{}).Build(), // TODO change when system inits genesis block and saves it will need to be read from db
+		mutex:        sync.RWMutex{},
+		merkle:       merkle,
+		layeredState: newLayeredState(persistence, int(config.StateStorageHistoryRetentionDistance())),
+		height:       height,
```

Comment:
but shouldn't `height` be a property of `layeredState`?

Reply:
moved this from service to layeredState (now renamed to rollingRevisions)